credoai/modules/model_modules/security.py

```diff
@@ -0,0 +1,223 @@
+        y_steal = self.y_test[indices[:len_steal]]
+        x_test = self.x_test[indices[len_steal:]]
+        y_test = self.y_test[indices[len_steal:]]
+
+        # extract
+        copycat = CopycatCNN(
+            classifier=self.victim_model,
+            nb_epochs=5,
+            nb_stolen=len_steal
+        )
+
+        thieved_model = self._get_model(x_steal.shape[1])
+        thieved_classifier = KerasClassifier(thieved_model)
+
+        thieved_classifier = copycat.extract(
+            x_steal, thieved_classifier=thieved_classifier)
+
+        # evaluate
+        y_true = [np.argmax(y, axis=None, out=None) for y in y_test]
+
+        y_pred = [np.argmax(y, axis=None, out=None)
+                  for y in thieved_classifier._model.predict(x_test)]
+        thieved_classifier_acc = sk_metrics.accuracy_score(y_true, y_pred)
+
+        y_pred = self.victim_model._model.predict(x_test)
+        victim_classifier_acc = sk_metrics.accuracy_score(y_true, y_pred)
+
+        metrics = {
+            'extraction_attack_score': max((thieved_classifier_acc - 0.5) / (victim_classifier_acc - 0.5), 0)
+        }
+
+        return metrics
+
+    def _get_model(self, input_dim):
+        """Creates a sequential binary classification model 
+
+        Parameters
+        ----------
+        input_dim : int
+            dimension of the feature vector
+        """
+        model = Sequential()
+        model.add(Dense(units=max(int(input_dim/2), 2),
+                  input_dim=input_dim, activation='relu'))
+        model.add(Dense(units=max(int(input_dim/4), 2), activation='relu'))
+        model.add(Dense(2, activation='sigmoid'))
+        model.compile(loss='binary_crossentropy',
+                      optimizer='adam', metrics=['accuracy'])
+
+        return model
+
+    def _evasion_attack(self, nsamples=100):
+        """Model evasion security attack
+
+        In model evasion, the adversary only has access to the prediction API of a target model
+            which she queries to extract information about the model internals and train a substitute model.
+
+        Returns
+        -------
+        dict
+            Key: evasion_attack_score
+            Value: area under success rate vs median distance threshold curve -- between 0 and 1
+        """
+        hsj = HopSkipJump(classifier=self.victim_model)
+
+        origl_sample = self.x_test[0:nsamples]
+        adver_sample = hsj.generate(origl_sample)
+
+        origl_pred = self.victim_model._model.predict(origl_sample)
+        adver_pred = self.victim_model._model.predict(adver_sample)
+
+        # standardize for robust distance calculation
+        scaler = StandardScaler()
+        scaler.fit(self.x_train)
+        origl_sample_scaled = scaler.transform(origl_sample)
+        adver_sample_scaled = scaler.transform(adver_sample)
+
+        # generate success rate vs median distance threshold curve
+        distances = np.diag(euclidean_distances(
```

Comment:
I changed this distance calculation to use numpy. It's faster and less error prone. There was also a "median" function that wasn't doing anything from what I could see. You'd calculate the norm between two vectors (which is a scalar) and then take the median. If I'm missing something please correct!

Reply:
Good catch. I did intend the median being there, but it seems it was not doing what it is supposed too. The idea is for the distance to be insensitive to the feature vector's length. Perhaps we can normalize the current distance by the length of the vector.